File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/session/ui/SessionViewHolder.java
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.session.ui;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v7.content.res.AppCompatResources;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import org.mozilla.focus.R;
import org.mozilla.focus.session.Session;
import org.mozilla.focus.session.SessionManager;
import org.mozilla.focus.telemetry.TelemetryWrapper;
import org.mozilla.focus.utils.UrlUtils;

import java.lang.ref.WeakReference;

public class SessionViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
    /* package */ static final int LAYOUT_ID = R.layout.item_session;

    private final SessionsSheetFragment fragment;
    private final TextView textView;
    private WeakReference<Session> sessionReference;

    /* package */ SessionViewHolder(SessionsSheetFragment fragment, View itemView) {
        super(itemView);

        this.fragment = fragment;

        textView = (TextView) itemView;
        textView.setOnClickListener(this);
    }

    /* package */ void bind(Session session) {
        this.sessionReference = new WeakReference<>(session);

        updateUrl(session);

        final boolean isCurrentSession = SessionManager.getInstance().isCurrentSession(session);
        final int actionColor = ContextCompat.getColor(textView.getContext(), R.color.colorAction);
        final int darkColor = ContextCompat.getColor(textView.getContext(), R.color.colorSession);

        updateTextColor(isCurrentSession, actionColor, darkColor);
        updateDrawable(isCurrentSession, actionColor, darkColor);
    }

    private void updateTextColor(boolean isCurrentSession, int actionColor, int darkColor) {
        textView.setTextColor(isCurrentSession ? actionColor : darkColor);
    }

    private void updateDrawable(boolean isCurrentSession, int actionColor, int darkColor) {
        final Drawable drawable = AppCompatResources.getDrawable(itemView.getContext(), R.drawable.ic_link);
        if (drawable == null) {
            return;
        }

        final Drawable wrapDrawable = DrawableCompat.wrap(drawable.mutate());
        DrawableCompat.setTint(wrapDrawable, isCurrentSession ? actionColor : darkColor);

        textView.setCompoundDrawablesWithIntrinsicBounds(wrapDrawable, null, null, null);
    }

    private void updateUrl(Session session) {
        final String shortUrl = UrlUtils.stripCommonSubdomains(
                UrlUtils.stripScheme(session.getUrl().getValue()));

        textView.setText(shortUrl);
    }

    @Override
    public void onClick(View view) {
        final Session session = sessionReference != null ? sessionReference.get() : null;
        if (session == null) {
            return;
        }

        selectSession(session);
    }

    private void selectSession(final Session session) {
        fragment.animateAndDismiss().addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                SessionManager.getInstance().selectSession(session);

                TelemetryWrapper.switchTabInTabsTrayEvent();
            }
        });
    }
}

File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/session/ui/SessionViewHolder.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.session.ui

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import androidx.recyclerview.widget.RecyclerView
import android.view.View
import android.widget.TextView
import mozilla.components.browser.session.Session
import org.mozilla.focus.R
import org.mozilla.focus.ext.beautifyUrl
import org.mozilla.focus.ext.requireComponents
import org.mozilla.focus.telemetry.TelemetryWrapper
import java.lang.ref.WeakReference

class SessionViewHolder internal constructor(
    private val fragment: SessionsSheetFragment,
    private val textView: TextView
) : RecyclerView.ViewHolder(textView), View.OnClickListener {
    companion object {
        @JvmField
        internal val LAYOUT_ID = R.layout.item_session
    }

    private var sessionReference: WeakReference<Session> = WeakReference<Session>(null)

    init {
        textView.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_link, 0, 0, 0)
        textView.setOnClickListener(this)
    }

    fun bind(session: Session) {
        this.sessionReference = WeakReference(session)

        updateTitle(session)

        val isCurrentSession = fragment.requireComponents.sessionManager.selectedSession == session

        updateTextBackgroundColor(isCurrentSession)
    }

    private fun updateTextBackgroundColor(isCurrentSession: Boolean) {
        val drawable = if (isCurrentSession) {
            R.drawable.background_list_item_current_session
        } else {
            R.drawable.background_list_item_session
        }
        textView.setBackgroundResource(drawable)
    }

    private fun updateTitle(session: Session) {
        textView.text =
            if (session.title.isEmpty()) session.url.beautifyUrl()
            else session.title
    }

    override fun onClick(view: View) {
        val session = sessionReference.get() ?: return
        selectSession(session)
    }

    private fun selectSession(session: Session) {
        fragment.animateAndDismiss().addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animation: Animator) {
                fragment.requireComponents.sessionManager.select(session)

                TelemetryWrapper.switchTabInTabsTrayEvent()
            }
        })
    }
}
